src/unitxt/type_utils.py

```diff
@@ -7,6 +7,114 @@
 from .utils import safe_eval
 
 
+def convert_union_type(type_string: str) -> str:
+    """Converts new Python union type hints (valid since Python 3.9) into form compatible with any Python version.
+
+    Args:
+        type_string (str): A string representation of a Python type hint. Examples include
+                           'List[int]', 'Dict[str, Any]', 'Optional[List[str]]', etc.
+
+    Returns:
+        str: A type string which is compatible with typing module. For example
+            'Union[int,float]' instead of 'int|float'.
+
```

Comment:
Hi @pawelknes , can you please formally describe the allowed input for your method, and also formally describe the generated output?

Is the following formal definition, using a mixture of regexes and Context Free Grammar,  in the direction of the input you have in mind?:

type -> basic | (typing.)?(List[type] | Dict[type, type] | Optional[type] | Tuple[type (, type)*] ) | type"|" type  ? 

basic ->  int, float, str, bool

Do you aim at all the "residents" of typing?  e.g. Literal, Annotated, Callable?

Reply:
To be honest I didn't consider remaining types but it should work anyway as long as a string represents a correct Python type hint (I'll extend documentation as you suggested). The function works based on positions of sings: `'[', ']', ','` so it won't matter if a dict or a literal is used. I also ran the following test and the output is correct:

`convert_union_type('Tuple[Annotated[int|float,ValueRange(-10, 5)],List[str|Literal["string"]],Callable[[int|float],[bool,str|List[str]]]]') -> 'Tuple[Annotated[Union[int,float],ValueRange(-10, 5)],List[Union[str,Literal["string"]]],Callable[[Union[int,float]],[bool,Union[str,List[str]]]]]'`

Reply:
What is the meaning of encapsulation by [ ]  "for no reason"?  I do not know..
I see in your example: `Callable[[int|float],[bool,str|List[str]]]`  what is the meaning of `[int | float]` ?  is it different from `int|float`  ?    Is this notatioin (encapsulating wihtin  [] ) allowed?